openedx_external_enrollments/external_enrollments.py

```diff
@@ -0,0 +1,424 @@
+                course_data["CourseName"] = salesforce_settings.get("Program_Name") or course.display_name
+                course_data["CourseCode"] = self._get_salesforce_course_id(course, course_id)
+                course_data["CourseStartDate"] = self._get_course_start_date(course, line.get("user_email"), course_id)
+                course_data["CourseEndDate"] = course.end.strftime("%Y-%m-%d")
+                course_data["CourseDuration"] = "0"
+            except:
+                pass
+            else:
+                courses.append(course_data)
+
+        return courses
+
+    def _get_salesforce_course_id(self, course, internal_id):
+        """
+        Returns either an external course id or internal.
+        """
+        if self._is_external_course(course):
+            return course.other_course_settings.get("external_course_run_id")
+
+        return internal_id
+
+    @staticmethod
+    def _is_external_course(course):
+        """
+        True if the course was confiured as external, False otherwise.
+        """
+
+        return (
+            course.other_course_settings.get("external_course_run_id") and
+            course.other_course_settings.get("external_course_target")
+        )
+
+    @staticmethod
+    def _get_course_start_date(course, email, course_id):
+        """
+        Return the course date start.
+        """
+
+        user, _ = get_user(email=email)
+        course_key = CourseKey.from_string(course_id)
+        enrollment = CourseEnrollment.get_enrollment(user, course_key)
+
+        if course.self_paced:
+            dates_to_check = [enrollment.created, course.start]
+            student_start = max(dates_to_check)
+        else:
+            student_start = course.start
+
+        return student_start.strftime("%Y-%m-%d")
+
+    def _get_enrollment_data(self, data, course_settings):
+        """
+        :param data:
+        :return:
+        """
+        valid_keys = [
+            "FirstName",
+            "LastName",
+            "Email",
+            "Company",
+            "Institution_Hidden",
+            "Type_Hidden",
+            "Program_of_Interest",
+            "Lead_Source",
+            "Secondary_Source",
+            "Tertiary_Source",
+            "Drupal_ID",
+            "Purchase_Type",
+            "PaymentAmount",
+            "Amount_Currency",
+            "Course_Data",
+        ]
+        payload = {
+            "enrollment": {}
+        }
+        openedx_user_info = self._get_openedx_user(data)
+        payload["enrollment"].update(openedx_user_info)
+
+        salesforce_data = self._get_salesforce_data(data)
+        payload["enrollment"].update(salesforce_data)
+
+        payload["enrollment"]["Course_Data"] = self._get_courses_data(
+            data,
+            data.get("supported_lines"),
+        )
+
+        unwanted = set(payload["enrollment"]) - set(valid_keys)
+        for unwanted_key in unwanted:
+            del payload["enrollment"][unwanted_key]
```

Comment:
Could a key error occur with the del statement? @andrey-canon

Reply:
out of the scope of this pr